File: chat/internal/services/notification_service.go
package services

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"kelisim-chat/internal/config"
	"kelisim-chat/internal/database"
	"kelisim-chat/internal/models"
	"net/http"
	"time"

	"github.com/sirupsen/logrus"
)

// NotificationService 推送通知服务
type NotificationService struct{}

// NewNotificationService 创建推送通知服务
func NewNotificationService() *NotificationService {
	return &NotificationService{}
}

// FCMLegacyMessage Firebase Cloud Messaging Legacy API 消息格式
type FCMLegacyMessage struct {
	To               string                 `json:"to"`
	Notification     FCMNotification        `json:"notification"`
	Data             map[string]interface{} `json:"data,omitempty"`
	Priority         string                 `json:"priority,omitempty"`
	ContentAvailable bool                   `json:"content_available,omitempty"`
}

type FCMNotification struct {
	Title string `json:"title"`
	Body  string `json:"body"`
}

// RegisterDeviceToken 注册设备 Token
func (s *NotificationService) RegisterDeviceToken(userID uint, token, platform, deviceID string) error {
	// 检查是否已存在
	var existingToken models.DeviceToken
	err := database.DB.Where("user_id = ? AND token = ?", userID, token).First(&existingToken).Error

	if err == nil {
		// 已存在，更新时间和状态
		return database.DB.Model(&existingToken).Updates(map[string]interface{}{
			"is_active":  true,
			"platform":   platform,
			"device_id":  deviceID,
			"updated_at": time.Now(),
		}).Error
	}

	// 不存在，创建新记录
	deviceToken := &models.DeviceToken{
		UserID:    userID,
		Token:     token,
		Platform:  platform,
		DeviceID:  deviceID,
		IsActive:  true,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}

	return database.DB.Create(deviceToken).Error
}

// UnregisterDeviceToken 注销设备 Token
func (s *NotificationService) UnregisterDeviceToken(userID uint, token string) error {
	return database.DB.Model(&models.DeviceToken{}).
		Where("user_id = ? AND token = ?", userID, token).
		Update("is_active", false).Error
}

// GetUserDeviceTokens 获取用户的所有活跃设备 Token
func (s *NotificationService) GetUserDeviceTokens(userID uint) ([]models.DeviceToken, error) {
	var tokens []models.DeviceToken
	err := database.DB.Where("user_id = ? AND is_active = ?", userID, true).Find(&tokens).Error
	return tokens, err
}

// SendNotificationToUser 发送通知给指定用户
func (s *NotificationService) SendNotificationToUser(userID uint, title, body string, data map[string]interface{}) error {
	// 获取用户的所有设备 Token
	tokens, err := s.GetUserDeviceTokens(userID)
	if err != nil {
		return err
	}

	if len(tokens) == 0 {
		logrus.Infof("User %d has no active device tokens", userID)
		return nil
	}

	// 发送推送通知到所有设备
	for _, deviceToken := range tokens {
		go func(token models.DeviceToken) {
			err := s.sendFCMNotification(token.Token, title, body, data, token.Platform)
			if err != nil {
				logrus.Errorf("Failed to send notification to device %s: %v", token.Token, err)
				// 如果是 token 无效，标记为不活跃
				if isInvalidTokenError(err) {
					s.UnregisterDeviceToken(token.UserID, token.Token)
				}
			}
		}(deviceToken)
	}

	return nil
}

// sendFCMNotification 通过 FCM Legacy API 发送推送通知
func (s *NotificationService) sendFCMNotification(token, title, body string, data map[string]interface{}, platform string) error {
	fcmServerKey := config.AppConfig.FCMServerKey
	if fcmServerKey == "" {
		logrus.Warn("FCM server key not configured")
		return nil
	}

	// 确保 data 中的值都是字符串（FCM Legacy API 要求）
	stringData := make(map[string]string)
	for k, v := range data {
		stringData[k] = fmt.Sprintf("%v", v)
	}

	// 构建 FCM Legacy 消息
	message := FCMLegacyMessage{
		To: token,
		Notification: FCMNotification{
			Title: title,
			Body:  body,
		},
		Data:             make(map[string]interface{}),
		Priority:         "high",
		ContentAvailable: true,
	}

	// 将字符串数据添加到消息中
	for k, v := range stringData {
		message.Data[k] = v
	}

	// 发送 HTTP 请求到 FCM
	jsonData, err := json.Marshal(message)
	if err != nil {
		return err
	}

	// 使用 Legacy FCM API (更简单，无需 OAuth)
	// 生产环境建议使用 FCM HTTP v1 API
	fcmURL := "https://fcm.googleapis.com/fcm/send"
	req, err := http.NewRequest("POST", fcmURL, bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("key=%s", fcmServerKey))

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("FCM returned status %d: %s", resp.StatusCode, string(body))
	}

	logrus.Infof("Push notification sent successfully to token: %s", token[:10]+"...")
	return nil
}

// SendChatMessageNotification 发送聊天消息通知
func (s *NotificationService) SendChatMessageNotification(message *models.Message, recipientUserIDs []uint) error {
	if message.Sender == nil {
		return fmt.Errorf("message sender is nil")
	}

	senderName := message.Sender.GetFullName()
	if senderName == "" {
		// Email 是指针类型，需要检查并解引用
		if message.Sender.Email != nil {
			senderName = *message.Sender.Email
		} else {
			senderName = "未知用户"
		}
	}

	title := fmt.Sprintf("来自 %s 的新消息", senderName)
	body := ""

	// 根据消息类型设置内容
	switch message.Type {
	case "text":
		if message.Content != nil {
			body = *message.Content
		}
	case "image":
		body = "[图片]"
	case "document":
		body = "[文档]"
	default:
		body = "[消息]"
	}

	// 构建数据
	data := map[string]interface{}{
		"type":       "chat_message",
		"chat_id":    fmt.Sprintf("%d", message.ChatID),
		"message_id": fmt.Sprintf("%d", message.ID),
		"sender_id":  fmt.Sprintf("%d", *message.SenderID),
	}

	// 发送给所有接收者
	for _, userID := range recipientUserIDs {
		// 不发送给发送者自己
		if message.SenderID != nil && userID == *message.SenderID {
			continue
		}

		go func(uid uint) {
			err := s.SendNotificationToUser(uid, title, body, data)
			if err != nil {
				logrus.Errorf("Failed to send notification to user %d: %v", uid, err)
			}
		}(userID)
	}

	return nil
}

File: laravel/app/Services/NotificationService.php
<?php

namespace App\Services;

use App\Enums\NotificationType;
use App\Models\Notification;
use App\Models\User;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;

class NotificationService
{
    /**
     * 创建通知
     *
     * @param User|int $user 用户或用户ID
     * @param string $type 通知类型
     * @param string $title 通知标题
     * @param string $content 通知内容
     * @param array $data 额外数据
     * @return Notification
     */
    public function createNotification($user, string $type, string $title, string $content, array $data = []): Notification
    {
        $userId = $user instanceof User ? $user->id : $user;

        $notification = Notification::create([
            'user_id' => $userId,
            'type' => $type,
            'title' => $title,
            'content' => $content,
            'data' => $data,
        ]);

        // Send push notification
        $this->sendPushNotification($userId, $title, $content, $data);

        return $notification;
    }

    /**
     * Send push notification via Firebase Cloud Messaging
     *
     * @param int $userId
     * @param string $title
     * @param string $body
     * @param array $data
     * @return void
     */
    protected function sendPushNotification(int $userId, string $title, string $body, array $data = []): void
    {
        $fcmServerKey = env('FCM_SERVER_KEY');
        
        if (empty($fcmServerKey)) {
            Log::warning('FCM_SERVER_KEY not configured, skipping push notification');
            return;
        }

        try {
            // Prepare payload
            $data = $this->preparePushData($data);

            // Get user device tokens from chat service database
            $deviceTokens = $this->getUserDeviceTokens($userId);
            
            if (empty($deviceTokens)) {
                Log::info("User {$userId} has no active device tokens");
                return;
            }

            // Send notification to each device
            foreach ($deviceTokens as $token) {
                $this->sendFCMNotification($token, $title, $body, $data, $fcmServerKey);
            }
        } catch (\Exception $e) {
            Log::error('Failed to send push notification: ' . $e->getMessage());
        }
    }

    /**
     * Get user device tokens from chat service database
     *
     * @param int $userId
     * @return array
     */
    protected function getUserDeviceTokens(int $userId): array
    {
        try {
            // Query device_tokens table directly
            $tokens = \DB::connection('mysql')
                ->table('device_tokens')
                ->where('user_id', $userId)
                ->where('is_active', true)
                ->pluck('token')
                ->toArray();
            
            return $tokens;
        } catch (\Exception $e) {
            Log::error('Failed to fetch device tokens: ' . $e->getMessage());
            return [];
        }
    }

    /**
     * Send FCM notification
     *
     * @param string $token
     * @param string $title
     * @param string $body
     * @param array $data
     * @param string $serverKey
     * @return void
     */
    protected function sendFCMNotification(string $token, string $title, string $body, array $data, string $serverKey): void
    {
        try {
            // Convert all data values to strings (FCM requirement)
            $stringData = [];
            foreach ($data as $key => $value) {
                $stringData[$key] = (string) $value;
            }

            $response = Http::withHeaders([
                'Authorization' => 'key=' . $serverKey,
                'Content-Type' => 'application/json',
            ])->post('https://fcm.googleapis.com/fcm/send', [
                'to' => $token,
                'notification' => [
                    'title' => $title,
                    'body' => $body,
                ],
                'data' => $stringData,
                'priority' => 'high',
                'content_available' => true,
            ]);

            if ($response->successful()) {
                Log::info("Push notification sent successfully to token: " . substr($token, 0, 10) . '...');
            } else {
                Log::error("FCM error: " . $response->body());
            }
        } catch (\Exception $e) {
            Log::error('Failed to send FCM notification: ' . $e->getMessage());
        }
    }

    /**
     * 创建咨询状态变更通知
     *
     * @param User|int $user 用户
     * @param int $consultationId 咨询ID
     * @param string $oldStatus 旧状态
     * @param string $newStatus 新状态
     * @return Notification
     */
    public function notifyConsultationStatusChange($user, int $consultationId, string $oldStatus, string $newStatus): Notification
    {
        $statusLabels = [
            'pending' => '待处理',
            'in_progress' => '处理中',
            'archived' => '已归档',
            'cancelled' => '已取消',
        ];

        $title = '咨询状态变更';
        $content = sprintf(
            '您的咨询状态已从「%s」变更为「%s」',
            $statusLabels[$oldStatus] ?? $oldStatus,
            $statusLabels[$newStatus] ?? $newStatus
        );

        $data = $this->injectTranslationMeta(
            NotificationType::CONSULTATION_STATUS,
            [
                'consultation_id' => $consultationId,
                'old_status' => $oldStatus,
                'new_status' => $newStatus,
            ],
            [
                'oldStatus' => $oldStatus,
                'newStatus' => $newStatus,
            ]
        );

        return $this->createNotification(
            $user,
            NotificationType::CONSULTATION_STATUS,
            $title,
            $content,
            $data
        );
    }

    /**
     * 创建咨询分配通知（律师被分配咨询）
     *
     * @param User|int $lawyer 律师
     * @param int $consultationId 咨询ID
     * @param string $consultationTitle 咨询标题
     * @return Notification
     */
    public function notifyConsultationAssigned($lawyer, int $consultationId, string $consultationTitle): Notification
    {
        $title = '新咨询分配';
        $content = sprintf('您收到了一个新的咨询：%s', $consultationTitle);

        $data = $this->injectTranslationMeta(
            NotificationType::CONSULTATION_ASSIGNMENT,
            [
                'consultation_id' => $consultationId,
                'consultation_title' => $consultationTitle,
            ],
            [
                'consultationTitle' => $consultationTitle,
            ]
        );

        return $this->createNotification(
            $lawyer,
            NotificationType::CONSULTATION_ASSIGNMENT,
            $title,
            $content,
            $data
        );
    }

    /**
     * 创建优先级提升通知
     *
     * @param User|int $lawyer 律师
     * @param int $consultationId 咨询ID
     * @param string $oldPriority 旧优先级
     * @param string $newPriority 新优先级
     * @return Notification
     */
    public function notifyPriorityEscalated($lawyer, int $consultationId, string $oldPriority, string $newPriority): Notification
    {
        $priorityLabels = [
            'low' => '低',
            'medium' => '中',
            'high' => '高',
            'urgent' => '紧急',
        ];

        $title = '咨询优先级提升';
        $content = sprintf(
            '咨询优先级已从「%s」提升至「%s」',
            $priorityLabels[$oldPriority] ?? $oldPriority,
            $priorityLabels[$newPriority] ?? $newPriority
        );

        $data = $this->injectTranslationMeta(
            NotificationType::CONSULTATION_PRIORITY,
            [
                'consultation_id' => $consultationId,
                'old_priority' => $oldPriority,
                'new_priority' => $newPriority,
            ],
            [
                'oldPriority' => $oldPriority,
                'newPriority' => $newPriority,
            ]
        );

        return $this->createNotification(
            $lawyer,
            NotificationType::CONSULTATION_PRIORITY,
            $title,
            $content,
            $data
        );
    }


    /**
     * 创建消息通知
     *
     * @param User|int $user 用户
     * @param int $chatId 聊天ID
     * @param string $senderName 发送者名称
     * @param string $messagePreview 消息预览
     * @return Notification
     */
    public function notifyNewMessage($user, int $chatId, string $senderName, string $messagePreview): Notification
    {
        $title = '新消息';
        $content = sprintf('%s: %s', $senderName, $messagePreview);

        $data = $this->injectTranslationMeta(
            NotificationType::MESSAGE,
            [
                'chat_id' => $chatId,
                'sender_name' => $senderName,
                'message_preview' => $messagePreview,
            ],
            [
                'sender' => $senderName,
                'message' => $messagePreview,
            ]
        );

        return $this->createNotification(
            $user,
            NotificationType::MESSAGE,
            $title,
            $content,
            $data
        );
    }

    /**
     * 创建系统通知
     *
     * @param User|int $user 用户
     * @param string $title 标题
     * @param string $content 内容
     * @param array $data 额外数据
     * @return Notification
     */
    public function notifySystem($user, string $title, string $content, array $data = []): Notification
    {
        return $this->createNotification(
            $user,
            NotificationType::SYSTEM,
            $title,
            $content,
            $data
        );
    }

    /**
     * 批量创建通知
     *
     * @param array $userIds 用户ID数组
     * @param string $type 通知类型
     * @param string $title 通知标题
     * @param string $content 通知内容
     * @param array $data 额外数据
     * @return int 创建的通知数量
     */
    public function createBulkNotifications(array $userIds, string $type, string $title, string $content, array $data = []): int
    {
        $notifications = [];
        $now = now();

        foreach ($userIds as $userId) {
            $notifications[] = [
                'user_id' => $userId,
                'type' => $type,
                'title' => $title,
                'content' => $content,
                'data' => json_encode($data),
                'is_read' => false,
                'created_at' => $now,
                'updated_at' => $now,
            ];
        }

        Notification::insert($notifications);

        return count($notifications);
    }

    /**
     * 创建公告通知（给所有用户）
     *
     * @param string $title 标题
     * @param string $content 内容
     * @param array $data 额外数据
     * @return int 创建的通知数量
     */
    public function createAnnouncementForAll(string $title, string $content, array $data = []): int
    {
        $userIds = User::where('status', User::STATUS_ACTIVE)->pluck('id')->toArray();

        return $this->createBulkNotifications(
            $userIds,
            NotificationType::ANNOUNCEMENT,
            $title,
            $content,
            $data
        );
    }

    /**
     * 为通知附加翻译元数据，供前端多语言渲染。
     */
    protected function injectTranslationMeta(string $type, array $data, array $params = []): array
    {
        $keys = NotificationType::translationKeys($type);

        if (!empty($keys['title']) || !empty($keys['body'])) {
            $data['translation'] = [
                'title_key' => $keys['title'],
                'body_key' => $keys['body'],
                'params' => $params,
            ];
        }

        return $data;
    }

    /**
     * 将通知数据转换为字符串，确保兼容 FCM payload 要求
     */
    protected function preparePushData(array $data): array
    {
        $normalized = [];

        foreach ($data as $key => $value) {
            if ($key === 'translation') {
                continue;
            }

            if (is_array($value) || is_object($value)) {
                $normalized[$key] = json_encode($value, JSON_UNESCAPED_UNICODE);
            } elseif (is_bool($value)) {
                $normalized[$key] = $value ? '1' : '0';
            } elseif ($value === null) {
                $normalized[$key] = '';
            } else {
                $normalized[$key] = $value;
            }
        }

        return $normalized;
    }
}


